rust_src/src/keymap.rs

```diff
@@ -673,4 +675,80 @@ pub fn copy_keymap(keymap: LispObject) -> LispObject {
     copy
 }
 
+// GC is possible in this funtion if it autoloads a keymap.
+/// Return the binding for command KEY in current keymaps.
+/// KEY is a string or vector, a sequence of keystrokes.
+/// The binding is probably a symbol with a function definition.
+///
+/// Normally, `key-binding' ignores bindings for t, which act as default
+/// bindings, used when nothing else in the keymap applies; this makes it
+/// usable as a general function for probing keymaps.  However, if the
+/// optional second argument ACCEPT-DEFAULT is non-nil, `key-binding' does
+/// recognize the default bindings, just as `read-key-sequence' does.
+///
+/// Like the normal command loop, `key-binding' will remap the command
+/// resulting from looking up KEY by looking up the command in the
+/// current keymaps.  However, if the optional third argument NO-REMAP
+/// is non-nil, `key-binding' returns the unmapped command.
+///
+/// If KEY is a key sequence initiated with the mouse, the used keymaps
+/// will depend on the clicked mouse position with regard to the buffer
+/// and possible local keymaps on strings.
+///
+/// If the optional argument POSITION is non-nil, it specifies a mouse
+/// position as returned by `event-start' and `event-end', and the lookup
+/// occurs in the keymaps associated with it instead of KEY.  It can also
+/// be a number or marker, in which case the keymap properties at the
+/// specified buffer position instead of point are used.
+#[lisp_fn(min = "1")]
+pub fn key_binding(
+    key: LispObject,
+    accept_default: bool,
+    no_remap: bool,
+    mut position: LispObject,
+) -> LispObject {
+    if key.is_vector() && position.is_nil() {
+        let key = key.force_vector();
+        if key.len() == 0 {
+            return Qnil;
+        }
+
+        // Mouse events may have a symbolic prefix indicating the scrollbar or modeline
+        let idx = if key.get(0).is_symbol() && key.len() > 1 {
+            1
+        } else {
+            0
+        };
+        let event: keyboard::Event = key.get(idx).into();
+
+        // Ignore locations without event data
+        if event.has_data() {
+            let kind = event.head_kind();
+            if kind == Qmouse_click {
+                position = event.start()
+            }
+        }
+    }
+
+    let value = lookup_key(
+        (Qkeymap, unsafe { Fcurrent_active_maps(Qt, position) }).into(),
+        key,
+        accept_default.into(),
+    );
+
+    if value.is_nil() || value.is_integer() {
+        return Qnil;
+    }
+
+    // If the result of the ordinary keymap lookup is an interactive
+    // command, look for a key binding (i.e. remapping) for that command.
+    if !no_remap && value.is_symbol() {
+        let value = unsafe { Fcommand_remapping(value, position, Qnil) };
+        if value.is_not_nil() {
+            return value;
```

Comment:
Would it make sense to use `match` to avoid the additional variable. If not, I would at least rename the variable. I think it's easier to read if we don't shadow the variable here.

Reply:
I don't think match is able to help here, but renaming the variable does make things a lot more clear.